File: library/src/main/java/com/jerrylee/lib/mvvm/BaseFragment.java
package com.jerrylee.lib.mvvm;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import butterknife.ButterKnife;
import butterknife.Unbinder;
import me.yokeyword.fragmentation.SupportFragment;

/**
 * Description:
 * Author: jerry Lee
 * Date: 2019/06/27 15:42
 */
public abstract class BaseFragment extends SupportFragment {
    private Unbinder binder;
    protected String TAG;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(getLayoutResId(),container,false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        TAG = getClass().getSimpleName();
        binder = ButterKnife.bind(this, view);
        initConfig(savedInstanceState);
        initUI(view,savedInstanceState);
        initData(savedInstanceState);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (binder != null)
            binder.unbind();
    }

    /**
     * 初始化UI
     */
    public abstract void initUI(View view, @Nullable Bundle savedInstanceState);

    protected void initData(Bundle savedInstanceState){}

    protected void initConfig(Bundle savedInstanceState){}

    /**
     * 处理回退事件
     *
     * @return true 事件已消费
     * <p>
     * false 事件向上传递
     */
    @Override
    public boolean onBackPressedSupport() {
        if (getFragmentManager().getBackStackEntryCount() > 1) {
            //如果当前存在fragment>1，当前fragment出栈
            pop();
        } else {
            //已经退栈到root fragment，交由Activity处理回退事件
            return false;
        }
        return true;
    }

    /**
     * @return
     */
    protected abstract int getLayoutResId();


}

File: library/src/main/java/com/jerrylee/lib/mvvm/BaseFragment.kt
package com.jerrylee.lib.mvvm

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.jerrylee.lib.widget.stateview.core.LoadManager

/**
 * Description:
 * Author: jerry Lee
 * Date: 2019/06/27 15:42
 */
abstract class BaseFragment : Fragment() {
    private lateinit var rootView: View

    protected var fragmentActivity:FragmentActivity?=null

    protected var loadManager: LoadManager? = null

    //懒加载
    private var isFirstVisible:Boolean = true

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        rootView = inflater.inflate(getLayoutResId(), null, false)
        val contentLayout:View? = rootView.findViewById(getContentId())
        //是否需要添加loadingView
        contentLayout?.let {
            loadManager = LoadManager.Builder()
                    .setViewParams(it)
                    .setListener { onStateRefresh() }
                    .build()
        }
        initUI(rootView, savedInstanceState)
        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initConfig(savedInstanceState)
        initData(savedInstanceState)

        val isVis = isHidden || userVisibleHint
        if (isVis && isFirstVisible) {
            lazyLoad()
            isFirstVisible = false
        }
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        if (!hidden) {
            onVisible()
        } else {
            onInVisible()
        }
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        if (isVisibleToUser) {
            onVisible()
        } else {
            onInVisible()
        }
    }

    abstract fun getLayoutResId():Int

    /**
     * 获取需要显示loadingView的视图id
     */
    protected open fun getContentId() = -1

    /**
     * 初始化UI
     */
    abstract fun initUI(view: View, savedInstanceState: Bundle?)

    protected open fun initData(savedInstanceState: Bundle?) {}

    protected open fun initConfig(savedInstanceState: Bundle?) {}

    protected open fun onStateRefresh() {}

    /**
     * 当界面可见时的操作
     */
    protected fun onVisible() {
        if (isFirstVisible && isResumed) {
            lazyLoad()
            isFirstVisible = false
        }
    }

    /**
     * 数据懒加载
     */
    protected open fun lazyLoad() {

    }

    /**
     * 当界面不可见时的操作
     */
    protected open fun onInVisible() {

    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        fragmentActivity = context as? FragmentActivity
    }

    override fun onDetach() {
        super.onDetach()
        fragmentActivity = null
    }





}
